_sips/sips/2017-10-25-adding-prefix-types.md

```diff
@@ -0,0 +1,163 @@
+
+```scala
+import singleton.ops._
+
+object PrefixExample {
+  /*
+  We would much rather write the following to acheive more clarity and shorter code:
+  type Foo[Cond1, Cond2, Num] = ITE[Cond1 && !Cond2, -Num, Num]	
+  */
+  type Foo[Cond1, Cond2, Num] = ITE[Cond1 && ![Cond2], Negate[Num], Num]  
+  def foo[Cond1, Cond2, Num](implicit f : Foo[Cond1, Cond2, Num]) : f.Out = f.value
+  def foo(cond1 : Boolean, cond2 : Boolean, num : Int) : Int = 
+    if (cond1 && !cond2) -num else num
+}
+
+import PrefixExample._
+
+foo[true, false, 3] //returns -3
+foo(true, false, 3) //returns -3
+```
+
+Note: `type ![A]` is possible to define, but `type -[A]` is not due to collision with infix type parsing.
+
+---
+
+## Implementation
+
+A PR for this SIP is available at: [https://github.com/scala/scala/pull/6148](https://github.com/scala/scala/pull/6148)
+
+------
+
+### Interactions with other language features
+
+#### Variance Annotation
+Variance annotation uses the `-` and `+` symbols to annotate contravariant and covariant subtyping, respectively. Introducing unary prefix types may lead to some developer confusion. However, such interaction is very unlikely to occur. E.g.:
+
+```scala
+trait Negate[A]
+trait Positive[A]
+type unary_-[A] = Negate[A]
+type unary_+[A] = Positive[A]
+trait Contravariant[B, -A <: +B] //contravariant A subtype upper-bounded by Positive[B]
+trait Covariant[B, +A <: -B] //covariant A subtype upper-bounded by Negative[B]
+```
+
+#### Negative Literal Types
+Negative literal types are annotated using the `-` symbol. This can lead to the following confusion:
+
+```scala
+trait Negate[A]
+type unary_-[A] = Negate[A]
+trait MyTrait[B]
+
+type MinusFortyTwo = MyTrait[-42]
+type NegateFortyTwo = MyTrait[Negate[42]]
+```
+
+The above example demonstrates a case of two types `MinusFortyTwo` and `NegateFortyTwo` which are different. They may be equivalent in view (implicit conversion between the two type instances), but they are not equal.
+
+Note: It is not possible to annotate a positive literal type in Scala (checked both in TLS and Dotty):
+
+```scala
+val a : 42 = +42 //works
+val b : -42 = -42 //works
+val c : +42 = 42 //error: ';' expected but integer literal found
+```
+
+This means that if unary prefix types are added, then `+42` will be a type expansion of `unary_+[42]`.
+
+**Related Issues**
+* [Dotty Issue #2783](https://github.com/lampepfl/dotty/issues/2783)
+* [Typelevel Scala Issue #157](https://github.com/typelevel/scala/issues/157)
+
+Both SIP23 implementation and Dotty's implementation of literal types currently fail compilation when infix types interact with a negative literal type.
+```scala
+type ~~[A, B]
+type good = 2 ~~ 2
+type bad = 2 ~~ -2 //Error:(9, 20) ';' expected but integer literal found.
+type work_around = 2 ~~ (-2) //works for Typelevel scala, but fails in Dotty
+```
+It is not yet clear if this is an implementation issue, or if the spec should be changed to allow this as well.
```

Comment:
I think it's important that this is sorted out before presenting the proposal to the Committee (you want to give the impression that you've researched and understood future interactions with prefix types). In particular, why is this happening?
  
You may also ask the Dotty team about more clarifications in the issue tracker if you cannot find the root of the problem. /cc @smarter

Reply:
After investigating this further for TLS, I figured out that the SIP23 proposal (and implementation) should be changed to support infix types and negative literal types. Interestingly, implementing prefix types resolves this issue, because a `-` prefix type has a special case to handle negative literal types. Here is a branch I made that fixes SIP23:
https://github.com/soronpo/scala/blob/3aeb19a7ff9e9c2a1dda61899d4c6482cdd8343e/src/compiler/scala/tools/nsc/ast/parser/Parsers.scala#L1052
You can see that a Spec change is required to handle `NegativeNumericLiteral` inside a compound type.
The prefix expressions do the same thing, so that is why you don't have such a bug with infix expression and negative literal values.

Reply:
I think that SIP-23 should handle negated numeric literal types more smoothly than it does at present (see the example at the head of https://github.com/typelevel/scala/issues/157). However I think that the prefix type proposal goes beyond and is somewhat orthogonal to this, so I think it deserves a separate SIP.

Reply:
@soronpo Can you update this last part and we merge? Thanks for your hard work Oron.